libs/traits/src/rewards.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright 2022 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use sp_arithmetic::traits::Unsigned;
+use sp_runtime::{
+	traits::{One, Zero},
+	ArithmeticError, DispatchError, DispatchResult, FixedPointNumber, FixedPointOperand,
+};
+use sp_std::vec::Vec;
+
+use crate::ops::ensure::{EnsureAdd, EnsureFixedPointNumber};
+
+/// Abstraction over a distribution reward groups.
+pub trait GroupRewards {
+	/// Type used as balance for all currencies and reward.
+	type Balance;
+
+	/// Type used to identify the group
+	type GroupId;
+
+	/// Reward a group distributing the reward amount proportionally to all associated accounts.
+	/// This method is called by distribution method only when the group has some stake.
+	fn reward_group(group_id: Self::GroupId, reward: Self::Balance) -> DispatchResult;
+
+	/// Retrieve the total staked amount.
+	fn group_stake(group_id: Self::GroupId) -> Self::Balance;
+}
+
+/// Distribution mechanisms over group rewards.
+/// This trait is implemented automatically for all `GroupRewards` with the requested bounds.
+pub trait DistributedRewards: GroupRewards
+where
+	<Self as GroupRewards>::Balance: FixedPointOperand + Zero,
+	<Self as GroupRewards>::GroupId: Clone,
+{
+	/// Distribute uniformly the reward given to the entire list of groups.
+	/// Only groups with stake will be taken for distribution.
+	///
+	/// This method makes several calls to `Rewards::reward_group()` under the hood.
+	/// If one of those calls fail, this method will continue to reward the rest of the groups,
+	/// The failed group errors will be returned.
+	fn distribute_reward<Rate, It>(
```

Comment:
I'm not sure if I understand well the drawback, but only remark that the caller can choose the groups they want to distribute each time. So if there is some group that is not ready (without the stake. i.e), the caller can omit it from the group_id list until it has some value.

Reply:
Putting a bit more thought to it the potential logic of the block_rewards, the `on_initialize` would have to iterate over the list of current collators and manually add entries into the stake[i] (or call `deposit_stake`) and then call this logic to calculate the rewards (with a single group & unity weight).
Otherwise, I do not know how we can have accumulation of rewards in a scalable manner.